src/main/java/org/jenkinsci/plugins/xunit/XUnitProcessor.java

```diff
@@ -275,32 +327,36 @@ private TestResultAction getPreviousTestResultAction(Run<?, ?> build) {
         return getTestResultAction(previousBuild);
     }
 
-    private void recordTestResult(Run<?, ?> build, FilePath workspace, TaskListener listener, XUnitLog xUnitLog) throws XUnitException {
-        TestResultAction existingAction = build.getAction(TestResultAction.class);
-        final long buildTime = build.getTimestamp().getTimeInMillis();
-        final long nowMaster = System.currentTimeMillis();
-
-        TestResult existingTestResults = null;
-        if (existingAction != null) {
-            existingTestResults = existingAction.getResult();
-        }
-
-        TestResult result = getTestResult(workspace, "**/TEST-*.xml", existingTestResults, buildTime, nowMaster);
-        if (result != null) {
-            TestResultAction action;
-            if (existingAction == null) {
+    private void recordTestResult(Run<?, ?> build, FilePath workspace, TaskListener listener, XUnitLog xUnitLog,
+                                  @CheckForNull String nodeId, List<String> enclosingBlocks) throws XUnitException {
+        synchronized (build) {
+            TestResultAction action = build.getAction(TestResultAction.class);
+            final long buildTime = build.getTimestamp().getTimeInMillis();
+            final long nowMaster = System.currentTimeMillis();
+
+            boolean appending = false;
+            TestResult result = getTestResult(workspace, "**/TEST-*.xml", null, buildTime, nowMaster,
+                    build.getExternalizableId(), nodeId, enclosingBlocks);
+            if (action == null) {
                 action = new TestResultAction(build, result, listener);
             } else {
-                action = existingAction;
-                action.setResult(result, listener);
+                appending = true;
+                result.freeze(action);
+                action.mergeResult(result, listener);
             }
 
             if (result.getPassCount() == 0 && result.getFailCount() == 0) {
                 xUnitLog.warningConsoleLogger("All test reports are empty.");
             }
 
-            if (existingAction == null) {
-                build.getActions().add(action);
+            if (appending) {
+                try {
+                    build.save();
+                } catch (IOException e) {
+                    throw new XUnitException("Error saving build", e);
+                }
+            } else {
+                build.addAction(action);
```

Comment:
Triggers automatic save?

Reply:
It's not: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/Run.java#L397-L404